x-pack/plugins/alerting/common/rule_task_instance.ts

```diff
@@ -9,7 +9,15 @@ import * as t from 'io-ts';
 import { rawAlertInstance } from './alert_instance';
 import { DateFromString } from './date_from_string';
 
+const actionSchema = t.partial({
+  group: t.string,
+  id: t.string,
+  actionTypeId: t.string,
+  params: t.record(t.string, t.unknown),
+});
+
 export const ruleStateSchema = t.partial({
+  triggeredActions: t.array(actionSchema),
```

Comment:
Do we need to persist this in the task state? This seems like it could unnecessarily bloat the task document fairly easily.

Here is what the task state looks like after an execution with 7 alerts with actions:

```
{
	"triggeredActions": [{
		"actionTypeId": ".server-log",
		"params": {
			"level": "info",
			"message": "alert 'test' is active for group 'host-5':\n\n- Value: 0.88900000055631\n- Conditions Met: avg(system.cpu.total.norm.pct) is greater than 0 over 1m\n- Timestamp: 2022-01-25T18:21:54.690Z"
		},
		"group": "threshold met",
		"id": "c0ad3e20-7e08-11ec-8a31-b505ec9f84d0"
	}, {
		"actionTypeId": ".server-log",
		"params": {
			"level": "info",
			"message": "alert 'test' is active for group 'host-3':\n\n- Value: 0.6550000014404456\n- Conditions Met: avg(system.cpu.total.norm.pct) is greater than 0 over 1m\n- Timestamp: 2022-01-25T18:21:54.690Z"
		},
		"group": "threshold met",
		"id": "c0ad3e20-7e08-11ec-8a31-b505ec9f84d0"
	}, {
		"actionTypeId": ".server-log",
		"params": {
			"level": "info",
			"message": "alert 'test' is active for group 'host-6':\n\n- Value: 0.6490000019470851\n- Conditions Met: avg(system.cpu.total.norm.pct) is greater than 0 over 1m\n- Timestamp: 2022-01-25T18:21:54.690Z"
		},
		"group": "threshold met",
		"id": "c0ad3e20-7e08-11ec-8a31-b505ec9f84d0"
	}, {
		"actionTypeId": ".server-log",
		"params": {
			"level": "info",
			"message": "alert 'test' is active for group 'host-2':\n\n- Value: 0.5293333355337382\n- Conditions Met: avg(system.cpu.total.norm.pct) is greater than 0 over 1m\n- Timestamp: 2022-01-25T18:21:54.690Z"
		},
		"group": "threshold met",
		"id": "c0ad3e20-7e08-11ec-8a31-b505ec9f84d0"
	}, {
		"actionTypeId": ".server-log",
		"params": {
			"level": "info",
			"message": "alert 'test' is active for group 'host-1':\n\n- Value: 0.4730000009139379\n- Conditions Met: avg(system.cpu.total.norm.pct) is greater than 0 over 1m\n- Timestamp: 2022-01-25T18:21:54.690Z"
		},
		"group": "threshold met",
		"id": "c0ad3e20-7e08-11ec-8a31-b505ec9f84d0"
	}, {
		"actionTypeId": ".server-log",
		"params": {
			"level": "info",
			"message": "alert 'test' is active for group 'host-4':\n\n- Value: 0.3259999992946784\n- Conditions Met: avg(system.cpu.total.norm.pct) is greater than 0 over 1m\n- Timestamp: 2022-01-25T18:21:54.690Z"
		},
		"group": "threshold met",
		"id": "c0ad3e20-7e08-11ec-8a31-b505ec9f84d0"
	}, {
		"actionTypeId": ".server-log",
		"params": {
			"level": "info",
			"message": "alert 'test' is active for group 'host-7':\n\n- Value: 0.25399999860674144\n- Conditions Met: avg(system.cpu.total.norm.pct) is greater than 0 over 1m\n- Timestamp: 2022-01-25T18:21:54.690Z"
		},
		"group": "threshold met",
		"id": "c0ad3e20-7e08-11ec-8a31-b505ec9f84d0"
	}],
	"alertInstances": {
		"host-5": {
			"state": {
				"start": "2022-01-25T18:21:54.790Z",
				"duration": 0
			},
			"meta": {
				"lastScheduledActions": {
					"group": "threshold met",
					"date": "2022-01-25T18:21:54.793Z"
				}
			}
		},
		"host-3": {
			"state": {
				"start": "2022-01-25T18:21:54.790Z",
				"duration": 0
			},
			"meta": {
				"lastScheduledActions": {
					"group": "threshold met",
					"date": "2022-01-25T18:21:54.794Z"
				}
			}
		},
		"host-6": {
			"state": {
				"start": "2022-01-25T18:21:54.790Z",
				"duration": 0
			},
			"meta": {
				"lastScheduledActions": {
					"group": "threshold met",
					"date": "2022-01-25T18:21:54.794Z"
				}
			}
		},
		"host-2": {
			"state": {
				"start": "2022-01-25T18:21:54.790Z",
				"duration": 0
			},
			"meta": {
				"lastScheduledActions": {
					"group": "threshold met",
					"date": "2022-01-25T18:21:54.795Z"
				}
			}
		},
		"host-1": {
			"state": {
				"start": "2022-01-25T18:21:54.790Z",
				"duration": 0
			},
			"meta": {
				"lastScheduledActions": {
					"group": "threshold met",
					"date": "2022-01-25T18:21:54.795Z"
				}
			}
		},
		"host-4": {
			"state": {
				"start": "2022-01-25T18:21:54.790Z",
				"duration": 0
			},
			"meta": {
				"lastScheduledActions": {
					"group": "threshold met",
					"date": "2022-01-25T18:21:54.796Z"
				}
			}
		},
		"host-7": {
			"state": {
				"start": "2022-01-25T18:21:54.790Z",
				"duration": 0
			},
			"meta": {
				"lastScheduledActions": {
					"group": "threshold met",
					"date": "2022-01-25T18:21:54.796Z"
				}
			}
		}
	},
	"previousStartedAt": "2022-01-25T18:21:54.486Z"
}
```

Reply:
Haven't actually reviewed yet, but yah, my first thought is that these don't need to be persisted in task state.  Which can get bloaty even without this (rules with a lot of state and/or alerts).

Reply:
I was forced by [this function](https://github.com/elastic/kibana/blob/57d507c12140fd491e5d60755c795d901c249d39/x-pack/plugins/alerting/server/lib/rule_execution_status.ts#L14) but i've been just told that we need to hold off the SO changes. So i am removing all those numberOfTriggeredActions fields from the state :)

Reply:
Aah sorry i have to keep that data somewhere in there.
What would you recommend? Maybe keeping in `alertTypeState` ?

Reply:
We could either 
1) Make `triggeredActions` optional on `RuleTaskState` and then omit it from the state that's returned at the end of the run() function in the alerting task runner (that state is the one that is persisted)

2) Be explicit and create a `RuleTaskStateWithActions` that includes `triggeredActions`. This would be what is returned from `loadRuleAttributesAndRun` and then what is returned from the run() function is still `RuleTaskState`.

Reply:
This complication points out that this module is in **dire** need of refactoring :-)

Ying's recommendations seem fine; probably an explicit type would be better, less chance for missing an optional property later.

I think I've tended to "pass things down" a lot in these functions, but I guess this is "coming back up" (return value, not parameter).  You could pass a mutable param, which would be "filled in" on the return.  Pretty icky, would not recommend.  However, perhaps there's some other mutable "thing" we're already passing down, that we could attach the data to.  My guess is ... that would be icky as well.